src/fs/inode_info.hh

```diff
@@ -0,0 +1,223 @@
+                --it;
+            }
+
+            while (it != data.end() and are_intersecting(range, it->second.data_range)) {
+                auto& data_vec = (it++)->second;
+                const auto cap = intersection(range, data_vec.data_range);
+                if (cap == data_vec.data_range) {
+                    // Fully intersects => execute
+                    execute_on_data_range_processor(data_vec.share_copy());
+                    continue;
+                }
+
+                inode_data_vec mid;
+                mid.data_range = std::move(cap);
+                auto mid_beg_shift = mid.data_range.beg - data_vec.data_range.beg;
+                std::visit(overloaded {
+                    [&](inode_data_vec::in_mem_data& mem) {
+                        mid.data_location = inode_data_vec::in_mem_data {
+                            mem.data.share(mid_beg_shift, mid.data_range.size())
+                        };
+                    },
+                    [&](inode_data_vec::on_disk_data& disk_data) {
+                        mid.data_location = inode_data_vec::on_disk_data {disk_data.device_offset + mid_beg_shift};
+                    },
+                    [&](inode_data_vec::hole_data&) {
+                        mid.data_location = inode_data_vec::hole_data {};
+                    },
+                }, data_vec.data_location);
+
+                // Execute on middle range
+                execute_on_data_range_processor(std::move(mid));
+            }
+        }
+    };
+
+    std::variant<directory, file> contents;
+
+    bool is_linked() const noexcept {
+        return directories_containing_file != 0;
+    }
+
+    bool is_open() const noexcept {
+        return opened_files_count != 0;
+    }
+
+    constexpr bool is_directory() const noexcept { return std::holds_alternative<directory>(contents); }
+
+    // These are noexcept because invalid access is a bug not an error
+    constexpr directory& get_directory() & noexcept { return std::get<directory>(contents); }
+    constexpr const directory& get_directory()  const & noexcept { return std::get<directory>(contents); }
+    constexpr directory&& get_directory() && noexcept { return std::move(std::get<directory>(contents)); }
+    constexpr const directory&& get_directory() const && noexcept { return std::move(std::get<directory>(contents)); }
+
+    constexpr bool is_file() const noexcept { return std::holds_alternative<file>(contents); }
+
+    // These are noexcept because invalid access is a bug not an error
+    constexpr file& get_file() & noexcept { return std::get<file>(contents); }
+    constexpr const file& get_file()  const & noexcept { return std::get<file>(contents); }
+    constexpr file&& get_file() && noexcept { return std::move(std::get<file>(contents)); }
+    constexpr const file&& get_file() const && noexcept { return std::move(std::get<file>(contents)); }
```

Comment:
Why would you ever need a const rvalue reference? I think these overloads just bloat the interface and are never actually useful. While at it, it's more human-readable when you omit the ampersand in `const & noexcept {` - this is the default and people are not used to seeing these.

To sum up, I'd rather `get_file` (and, respectively, `get_directory`) was declared as:
```cpp
    file& get_file() & noexcept { return std::get<file>(contents); }
    const file& get_file() const noexcept { return std::get<file>(contents); }
    file&& get_file() && noexcept { return std::move(std::get<file>(contents)); }
```
If you prefer (or need) to explicitly declare any of these as constexpr then I'm fine with that, but they can be safely omitted as well.

Reply:
This gives compilation error:
```cpp

struct bla {
	void foo() & {}
	void foo() const {}
	void foo() && {}
};

int main() {}
```
```gcc
/tmp/a.cc:3:7: error: ‘void bla::foo() const’ cannot be overloaded with ‘void bla::foo() &’
    3 |  void foo() const {}
      |       ^~~
/tmp/a.cc:2:7: note: previous declaration ‘void bla::foo() &’
    2 |  void foo() & {}
      |       ^~~
```

Reply:
Ok, you're right. `&` stays, so please just get rid of the unnecessary `const file&&` overload.